libs/datagrid/src/lib/ag-grid/use-column-sizes.ts

```diff
@@ -0,0 +1,138 @@
+import type { MutableRefObject, ReactElement } from 'react';
+import { useCallback, useState } from 'react';
+import type { AgGridReactProps, AgReactUiProps } from 'ag-grid-react';
+import type { Column, ColDef } from 'ag-grid-community';
+import debounce from 'lodash/debounce';
+import { create } from 'zustand';
+import { persist } from 'zustand/middleware';
+import { immer } from 'zustand/middleware/immer';
+import { useResizeObserver } from '@vegaprotocol/react-helpers';
+
+const STORAGE_KEY = 'vega_columns_sizes_store';
+const COLUMNS_SET_DEBOUNCE_TIME = 300;
+
+export const useColumnSizesStore = create<{
+  sizes: Record<string, Record<string, number>>;
+  valueSetter: (id: string, value: Record<string, number>) => void;
+}>()(
+  persist(
+    immer((set) => ({
+      sizes: {},
+      valueSetter: (id, value) =>
+        set((state) => {
+          state.sizes[id] = {
+            ...(state.sizes[id] || {}),
+            ...value,
+          };
+          return state;
+        }),
+    })),
+    { name: STORAGE_KEY }
+  )
+);
+
+interface UseColumnSizesProps {
+  id?: string;
+  container?: MutableRefObject<HTMLDivElement | null>;
+}
+export const useColumnSizes = ({
+  id = '',
+  container,
+}: UseColumnSizesProps): [
+  (columns: Column[]) => void,
+  (children?: ReactElement[]) => ReactElement[] | undefined,
+  (
+    props: AgGridReactProps | AgReactUiProps,
+    children?: ReactElement[]
+  ) => AgGridReactProps | AgReactUiProps
+] => {
+  const sizes = useColumnSizesStore((store) => store.sizes[id] || {});
+  const valueSetter = useColumnSizesStore((store) => store.valueSetter);
+  const getWidthOfAll = useCallback(
+    () =>
+      (container?.current as HTMLDivElement)?.getBoundingClientRect().width ??
+      0,
+    [container]
+  );
+  const recalculateSizes = useCallback(
+    (sizes: Record<string, number>) => {
+      const width = getWidthOfAll();
+      if (width && sizes['width'] && width !== sizes['width']) {
+        const oldWidth = sizes['width'];
+        const ratio = width / oldWidth;
+        return {
+          ...Object.entries(sizes).reduce((agg, [key, value]) => {
+            agg[key] = value * ratio;
+            return agg;
+          }, {} as Record<string, number>),
+          width,
+        } as Record<string, number>;
+      }
+      return sizes;
+    },
+    [getWidthOfAll]
+  );
+  const [calculatedSizes, setCalculatedSizes] = useState(
+    recalculateSizes(sizes)
+  );
+  const onResize = useCallback(() => {
+    const width = getWidthOfAll();
+    if (width && sizes['width'] && width !== sizes['width']) {
+      setCalculatedSizes(recalculateSizes(sizes));
+    }
+  }, [getWidthOfAll, recalculateSizes, sizes]);
+  useResizeObserver(container?.current as Element, onResize);
+
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const handleOnChange = useCallback(
+    debounce((columns: Column[]) => {
```

Comment:
chore (blocking): you can remove debounced logic as AgGrid provides a property event.finished which will be true once dragging is finished

Reply:
payload of the call when the `event.finished` is fired is much poorer than in other calls (with no `event.finished` = true). When event.finished is true, there is only one column dimension passed. Otherwise we have more siblings column, we can save simultaneously. I thought it can be more appropriate in our case.

Reply:
`event.columnApi.getColumnState()` does the trick